manuals/pffinfo.1
.Dd June 25, 2026
.Dt PFFINFO 1
.Os
.Sh NAME
.Nm pffinfo
.Nd determines information about a Personal Folder File (OST, PAB and PST)
.Sh SYNOPSIS
.Nm pffinfo
.Op Fl c Ar codepage
.Op Fl ahvV
.Ar source
.Sh DESCRIPTION
.Nm pffinfo
is a utility to determine information about a Personal Folder File (OST, PAB \
and PST)
.Pp
.Nm pffinfo
is part of the
.Nm libpff
package.
.Nm libpff
is a library to access the Personal Folder File (OST, PAB and PST) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl a
shows allocation information
.It Fl c Ar codepage
codepage of ASCII strings, options: ascii, windows-874, windows-932, \
windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 \
(default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 \
or windows-1258
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# pffinfo -a file.pst
pffinfo 20260625
.sp
Personal Folder File information:
	File size:		271360 bytes
	File content type:	Personal Storage Tables (PST)
	File type:		64-bit
	Encryption type:	compressible
.sp
Message store:
	Folders:		Subtree, Wastbox, Finder
	Password checksum:	N/A
.sp
Unallocated index node blocks:
	00018944 (0x00004a00) - 00020480 (0x00005000) size: 1536
	01051648 (0x00100c00) - 01052160 (0x00100e00) size: 512
.sp
Unallocated data blocks:
	00025024 (0x000061c0) - 00025536 (0x000063c0) size: 512
	01353664 (0x0014a7c0) - 01541120 (0x00178400) size: 187456
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr pffexport 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libpff/issues
.Sh COPYRIGHT
Copyright (C) 2008-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
libpff - clean up and refactoring

* refactor
** merge descriptors_index_tree and offsets_index_tree into index_tree?
** recover refactor libpff_recover_index_nodes to use index_node
** cache index nodes and/or index values
** cache local descriptors nodes and/or local descriptor values
** use or remove table->local_descriptors_cache

* libpff_table
  - move table header read functions into table header source file
  - move column definition read functions into column definition source file
  - cleaner reusable data structure? (libfdata)
  - create 7c table read entry definitions function
  - create ac table read entry definitions function
* have multi type work with data reference (in libpff_item) ?
* tree type implement use of initialize function (grep sizeof( libpff_tree_node_t ) )
* refactor direct use of array, list, and tree value attributes
* improve force decryption check
* refactor
  - create generic solution for data block handling for: individual blocks,
    array blocks, local descriptor blocks, table blocks
  - refactor libpff_local_descriptor_node_read to deal with compressed data
  blocks. read from data block?
    - fix libpff_local_descriptors_read_node_entry
  - item tree: wrap direct access to descriptors_index
  - recover: wrap direct access to descriptors and offsets_index
  - implement libfdata_list_set_element_by_index_with_mapped_size
  - 4k the back pointer of the array data block is the data identifier
    not the identifier?
    check for: "libpff_data_block_read: mismatch in data identifier:"
  - add a generic method to read data blocks from the file
  - local descriptors: cache local descriptor nodes ?
  - fix libfdata_list_set_element_data
  - table refactor entries into stand-alone multi set array ?
  - table read should empty the record_sets_array on read ?
  - record entry handle different data types on demand ?
  - what about table block?
  - use fvalue to store record entry data ?
  - libpff_item_values_get_record_entry_by_type add generic STRING TYPE ?
* code clean up
  - refactor table index into fdata list? with table blocks
  - does table index contain full copy of table data ? if so refactor
* refactor libpff_message_get_entry_value_utf8_string to match naming schema

* add threads RW lock to item
* remove libpff_data_block_clone
* pypff fix get_root_item

IOError: pypff_file_get_item_type_object: unable to retrieve item type.
libpff_table_read: invalid table: 0 - missing data identifier.
libpff_item_values_read: unable to read table.
libpff_item_values_get_record_entry_by_type: unable to read item values.
libpff_item_determine_type: unable to retrieve record entry: 0x001a.
libpff_item_get_type: unable to determine item type.

* add specific entry value flag for string types?
* add libpff_item_values_get_record_entry_by_type specific for string types?
* API add
  - libpff_multi_value_get_value_16bit
  - libpff_multi_value_get_value_floating_point
  - rename sub message to message ?
libpff
  - have libpff_record_entry_get_entry_multi_value use libfvalue data handle
    instead?
* pypff
  - create attachment item
  - make sure record set index and record entry index are uint32
  - add definitions
  - add record entry value functions
    pypff_record_entry_get_value_data_size
    get_value_data_as_boolean
    get_value_data_as_datetime
  - add multi value functions
  - add iterator objects (+/-)
  - add time as integer functions
  - add function to return offset of item
  - add more functions
  - add name to id map support
  - add message get_attachments_items function
  - add message get_recipients_items function

* local descriptor node caching
  - local descriptor node cache on top of data block cache ?
  - add cache timestamp
* add IPM.Post support (more1/Me.pst)
* corrupted
  - mark file with incorrect encryption as corrupted
* recovery
  - read data block size while scanning, ignore mismatches
  - check if local descriptors are read as block
  - handle corrupt file header and scan file?
* debug print
  - free map
  - densitiy list
* 64-bit 4k pages support
  - libpff_free_map.c
  - did the table value reference change?
* fix/check
  - check creation of .export when -t is a directory
  - remove copy file descriptors
  - Change most proxy pointers to const ?
  - check if libpff_item_get_entry_value_by_utf8_name works
  - Add function to easily map a named entry type to it's numeric value
  - add libpff_message_get_number_of_recipients() convenience function?
* add get entry id by name function
* add by_name functions (check corresponding TODO)
* fdata stream rewrite
  - array uses it as list
* clean up error path of libpff_item_tree_create_node

* libfdata update
  - improve stream, how deal with compressed data
  - fix libfdata_tree_initialize
* libpff IO handle
  - move index_nodes_vector and index_nodes_cache to file
* libfmapi
  - add debug output for PidTagFlagStatus
  - add debug output for PidLidSideEffects
- add debug output for PidLidTaskMode
- add debug output for PidLidAddressBookProviderArrayType
* implement libcdata support
  - refactor list
  - move tree node get value into libpff_item_initialize
* pffexport/export handle:
  - handle MIME attachments
  - protect against duplicate name clashes
  - dump attachment item values
  - add support for LIBPFF_ITEM_TYPE_COMMON (IPM)
  - separate item file into separate structure
  - move message specific funtionality into message file ?
  - allow to set output formatting utf8 or utf16
* item file:
  - create function to write enum values to the item file
  - create function to write flag values to the item file
* code clean up
  - what's the status of table block ?
  - unused and goto
* libpff: UTF8/UTF16
  - libpff_message_get_plain_text_body_size
  - libpff_message_get_rtf_body

* what about debug export and error notifications

Change item API into table API
* deprecrate and replace by copy data functions:
  - libpff_item_get_entry_value
  - libpff_item_get_entry_value_by_utf8_name

pfftools/pffinfo.c
/*
 * Shows information obtained from a Personal Folder File (OST, PAB and PST).
 *
 * Copyright (C) 2008-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "pfftools_getopt.h"
#include "pfftools_libcerror.h"
#include "pfftools_libclocale.h"
#include "pfftools_libcnotify.h"
#include "pfftools_libpff.h"
#include "pfftools_output.h"
#include "pfftools_signal.h"
#include "pfftools_unused.h"

info_handle_t *pffinfo_info_handle = NULL;
int pffinfo_abort                  = 0;

/* Signal handler for pffinfo
 */
void pffinfo_signal_handler(
      pfftools_signal_t signal PFFTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "pffinfo_signal_handler";

	PFFTOOLS_UNREFERENCED_PARAMETER( signal )

	pffinfo_abort = 1;

	if( pffinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     pffinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description    = \
		"Use pffinfo to determine information about a Personal Folder File (OST, PAB and PST).";

	pfftools_option_t options[ ] = {
		{ 'a', NULL, "shows allocation information" },
		{ 'c', "codepage", "codepage of ASCII strings, options: ascii, windows-874, windows-932, windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 (default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258" },
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libpff_error_t *error               = NULL;
	system_character_t *option_codepage = NULL;
	system_character_t *source          = NULL;
	char *program                       = "pffinfo";
	system_integer_t option             = 0;
	uint8_t show_allocation_information = 0;
	int number_of_options               = (int) ( sizeof( options ) / sizeof( pfftools_option_t ) );
	int result                          = 0;
	int verbose                         = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "pfftools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( pfftools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	pfftools_output_version_fprint(
	 stdout,
	 program );

	if( pfftools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = pfftools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				pfftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'a':
				show_allocation_information = 1;

				break;

			case (system_integer_t) 'c':
				option_codepage = optarg;

				break;

			case (system_integer_t) 'h':
				pfftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				pfftools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		pfftools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libpff_notify_set_stream(
	 stderr,
	 NULL );
	libpff_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &pffinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( option_codepage != NULL )
	{
		result = info_handle_set_ascii_codepage(
		          pffinfo_info_handle,
		          option_codepage,
		          &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set ASCII codepage in info handle.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported ASCII codepage defaulting to: windows-1252.\n" );
		}
	}
/* TODO
	if( pfftools_signal_attach(
	     pffinfo_signal_handler,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to attach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
*/
	if( info_handle_open_input(
	     pffinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	if( info_handle_file_fprint(
	     pffinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print file information.\n" );

		goto on_error;
	}
	if( info_handle_message_store_fprint(
	     pffinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print file stream and storage items.\n" );

		goto on_error;
	}
	if( show_allocation_information != 0 )
	{
		if( info_handle_unallocated_blocks_fprint(
		     pffinfo_info_handle,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to print file unallocated blocks.\n" );

			goto on_error;
		}
	}
/* TODO
	if( pfftools_signal_detach(
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to detach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
*/
	if( info_handle_close(
	     pffinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &pffinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( pffinfo_info_handle != NULL )
	{
		info_handle_free(
		 &pffinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

